packages/flutter/lib/src/widgets/nested_scroll_view.dart

```diff
@@ -0,0 +1,985 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/physics.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'notification_listener.dart';
+import 'primary_scroll_controller.dart';
+import 'scroll_activity.dart';
+import 'scroll_controller.dart';
+import 'scroll_interfaces.dart';
+import 'scroll_notification.dart';
+import 'scroll_physics.dart';
+import 'scroll_position.dart';
+import 'scroll_view.dart';
+import 'sliver.dart';
+import 'ticker_provider.dart';
+
+typedef ScrollActivity _NestedScrollActivityGetter(_NestedScrollPosition position);
+
+typedef List<Widget> NestedScrollViewOuterSliversBuilder(BuildContext context, bool innerBoxIsScrolled);
+
+class NestedScrollView extends StatefulWidget {
+  NestedScrollView({
+    Key key,
+    this.scrollDirection: Axis.vertical,
+    this.reverse: false,
+    this.physics,
+    @required this.outerSliverBuilder,
+    @required this.innerBox,
+  }) : super(key: key) {
+    assert(scrollDirection != null);
+    assert(reverse != null);
+    assert(outerSliverBuilder != null);
+    assert(innerBox != null);
+  }
+
+  // TODO(ianh): we should expose a controller so you can call animateTo, etc.
+
+  final Axis scrollDirection;
+
+  final bool reverse;
+
+  final ScrollPhysics physics;
+
+  final NestedScrollViewOuterSliversBuilder outerSliverBuilder;
+
+  final Widget innerBox;
+
+  double get initialScrollOffset => 0.0;
+
+  @protected
+  List<Widget> buildSlivers(BuildContext context, ScrollController innerController, bool innerBoxIsScrolled) {
+    final List<Widget> slivers = <Widget>[];
+    slivers.addAll(outerSliverBuilder(context, innerBoxIsScrolled));
+    slivers.add(new SliverFillRemaining(
+      child: new PrimaryScrollController(
```

Comment:
Needs documentation that horizontally scrolling children need to set primary explicitly to pick this up.